### ISSUE TITLE
proxy_cache_path not honoring max_size or inactive parameters

### ISSUE THREAD

**Issue by avtakkar** [nginx/nginx#282]
### Environment

Include the result of the following commands:
  - `nginx -V`

`nginx -V
nginx version: openresty/1.13.6.2
built by gcc 11.2.0 (GCC)
built with OpenSSL 3.0.3 3 May 2022
TLS SNI support enabled
configure arguments: --prefix=/usr/local/openresty/nginx --with-cc-opt=-O2 --add-module=../ngx_devel_kit-0.3.0 --add-module=../echo-nginx-module-0.61 --add-module=../xss-nginx-module-0.06 --add-module=../ngx_coolkit-0.2rc3 --add-module=../set-misc-nginx-module-0.32 --add-module=../form-input-nginx-module-0.12 --add-module=../encrypted-session-nginx-module-0.08 --add-module=../srcache-nginx-module-0.31 --add-module=../ngx_lua-0.10.13 --add-module=../ngx_lua_upstream-0.07 --add-module=../headers-more-nginx-module-0.33 --add-module=../array-var-nginx-module-0.05 --add-module=../memc-nginx-module-0.19 --add-module=../redis2-nginx-module-0.15 --add-module=../redis-nginx-module-0.3.7 --add-module=../rds-json-nginx-module-0.15 --add-module=../rds-csv-nginx-module-0.09 --add-module=../ngx_stream_lua-0.0.5 --with-ld-opt=-Wl,-rpath,/usr/local/openresty/luajit/lib --with-openssl=/tmp/openssl-3.0.3 --with-pcre=/tmp/pcre-8.42 --with-file-aio --with-http_addition_module --with-http_auth_request_module --with-http_dav_module --with-http_flv_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_image_filter_module=dynamic --with-http_mp4_module --with-http_random_index_module --with-http_realip_module --with-http_secure_link_module --with-http_slice_module --with-http_ssl_module --with-http_stub_status_module --with-http_sub_module --with-http_v2_module --with-http_xslt_module=dynamic --with-ipv6 --with-mail --with-mail_ssl_module --with-md5-asm --with-pcre-jit --with-sha1-asm --with-stream --with-stream_ssl_module --with-threads --with-stream --with-stream_ssl_module`

  - `uname -a`

`Linux ... 5.15.164.1-1.cm2 #1 SMP Sun Aug 18 19:16:21 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux`

### Description

The http file cache directory `/var/blobs/cache` is almost `14G` in size, and all data is expired (assuming the default `inactive` of 10 minutes and given that there is no traffic to the server anymore.)

Yet, the cache data is not cleared. I've tried restarting nginx and the VM it's running on as well. Pasted the debug logs I collected in the gist below after restarting nginx. 

#### nginx configuration
```
http
...
  proxy_cache_path /var/blobs/cache keys_zone=blobcache:30m max_size=10g;`
  ...
  server {
    ...
    localtion ... {
      ...
      proxy_cache blobcache;
      ...
    }
  }
...
```
#### nginx debug log

https://gist.github.com/avtakkar/b38a504df8204d7a5c86ccbe3cb202fd

**Comment by jo-carter** [comment]
@avtakkar NGINX performs lazy removal of cache entries on disk. From the [docs](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_path):

> The special “cache manager” process monitors the maximum cache size set by the max_size parameter, and the minimum amount of free space set by the min_free (1.19.1) parameter on the file system with cache. When the size is exceeded or there is not enough free space, it removes the least recently used data. The data is removed in iterations configured by manager_files, manager_threshold, and manager_sleep parameters (1.11.5).

If you set _min_free_ to the same value as _max_size,_ the cache entries should be deleted from disk as soon as expired.

**Comment by nmentz** [comment]
It looks like this is not vanilla nginx, but instead an OpenResty version that is based on nginx v1.13.6 (this version of nginx was released on Oct 10th 2017). The current release version of nginx sits at 1.27.1. The functionality of the nginx proxy module lies within src/http/modules/ngx_http_proxy_module.c. This file has seen significant changes from versions 1.13.6 to 1.27.1 as they are almost exactly eight years apart.

OpenResty is very good at closely following the nginx's release versions, I would take advantage of this and update the server from OpenResty's 1.13.6.2 to 1.27.1.1. If the issue persists, the opensource community will have a much easier time identifying the source of the bug.

**Comment by HanadaLee** [comment]
Please note that 10 minutes is just the minimum limit of inactivity, the cache manager will only delete when the allocated space is exceeded or the storage space is insufficient. This deletion is not greedy, it is deleted on demand, so you can't expect it to help you clear all files that have not been used for more than 10 minutes.

Here is the documentation for the cache manager
https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_path

> The special “cache manager” process monitors the maximum cache size set by the max_size parameter, and the minimum amount of free space set by the min_free (1.19.1) parameter on the file system with cache. When the size is exceeded or there is not enough free space, it removes the least recently used data. The data is removed in iterations configured by manager_files, manager_threshold, and manager_sleep parameters (1.11.5). During one iteration no more than manager_files items are deleted (by default, 100). The duration of one iteration is limited by the manager_threshold parameter (by default, 200 milliseconds). Between iterations, a pause configured by the manager_sleep parameter (by default, 50 milliseconds) is made.

**Comment by avtakkar** [comment]
> This file has seen significant changes from versions 1.13.6 to 1.27.1 as they are almost exactly eight years apart.

Yes, I'll try to update on our end, and see if the behaviour reproduces. I was wondering if there are any known issues that could be causing this.

> This deletion is not greedy, it is deleted on demand, so you can't expect it to help you clear all files that have not been used for more than 10 minutes.

Interesting. On most other machines, I have seen it clear the files very shortly after the configured thresholds are exceeded. Good to know, thanks.

**Comment by fabiorauber** [comment]
I'm seeing this issue in two of my servers running nginx/1.21.6. The cache is configured to 30g, and I had to manually delete the cache data to free up space. One server had 46GB of cache files and the other one, 53GB.

`nginx version: nginx/1.21.6
built by gcc 10.2.1 20210110 (Debian 10.2.1-6) 
built with OpenSSL 1.1.1n  15 Mar 2022
TLS SNI support enabled
configure arguments: --prefix=/usr/local/nginx --conf-path=/etc/nginx/nginx.conf --modules-path=/etc/nginx/modules --http-log-path=/var/log/nginx/access.log --error-log-path=/var/log/nginx/error.log --lock-path=/var/lock/nginx.lock --pid-path=/run/nginx.pid --http-client-body-temp-path=/var/lib/nginx/body --http-fastcgi-temp-path=/var/lib/nginx/fastcgi --http-proxy-temp-path=/var/lib/nginx/proxy --http-scgi-temp-path=/var/lib/nginx/scgi --http-uwsgi-temp-path=/var/lib/nginx/uwsgi --with-debug --with-compat --with-pcre-jit --with-http_ssl_module --with-http_stub_status_module --with-http_realip_module --with-http_auth_request_module --with-http_addition_module --with-http_geoip_module --with-http_gzip_static_module --with-http_sub_module --with-http_v2_module --with-stream --with-stream_ssl_module --with-stream_realip_module --with-stream_ssl_preread_module --with-threads --with-http_secure_link_module --with-http_gunzip_module --with-file-aio --without-mail_pop3_module --without-mail_smtp_module --without-mail_imap_module --without-http_uwsgi_module --without-http_scgi_module --with-cc-opt='-g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -Wno-deprecated-declarations -fno-strict-aliasing -D_FORTIFY_SOURCE=2 --param=ssp-buffer-size=4 -DTCP_FASTOPEN=23 -fPIC -I/root/.hunter/_Base/d45d77d/2761cc9/3b7ee27/Install/include -Wno-cast-function-type -m64 -mtune=generic' --with-ld-opt='-fPIE -fPIC -pie -Wl,-z,relro -Wl,-z,now -L/root/.hunter/_Base/d45d77d/2761cc9/3b7ee27/Install/lib' --user=www-data --group=www-data --add-module=/tmp/build/ngx_devel_kit-0.3.1 --add-module=/tmp/build/set-misc-nginx-module-0.33 --add-module=/tmp/build/headers-more-nginx-module-0.33 --add-module=/tmp/build/ngx_http_substitutions_filter_module-b8a71eacc7f986ba091282ab8b1bbbc6ae1807e0 --add-module=/tmp/build/lua-nginx-module-0.10.21 --add-module=/tmp/build/stream-lua-nginx-module-0.0.11 --add-module=/tmp/build/lua-upstream-nginx-module-8aa93ead98ba2060d4efd594ae33a35d153589bf --add-module=/tmp/build/nginx_ajp_module-fcbb2ccca4901d317ecd7a9dabb3fec9378ff40f --add-dynamic-module=/tmp/build/nginx-http-auth-digest-1.0.0 --add-dynamic-module=/tmp/build/nginx-opentracing-0.19.0/opentracing --add-dynamic-module=/tmp/build/ModSecurity-nginx-1.0.3 --add-dynamic-module=/tmp/build/ngx_http_geoip2_module-a26c6beed77e81553686852dceb6c7fdacc5970d --add-dynamic-module=/tmp/build/ngx_brotli`
